### PR TITLE
feat(builder): Describe optional columnList arg to with/Recurisve [knex/knex#4514]

### PR DESCRIPTION
**Pull request by jeremy-w**
These are the documentation updates to accompany https://github.com/knex/knex/pull/4652, to resolve knex/knex#4514.

I'm not 100% I got the notation right; I'm not clear on whether it's `[whatever]` to signal an array vs `[whatever]` to signal an optional argument. But I hope the examples make the meaning clear.

### CODE REVIEW

**Comment by kibertoad** [comment]
@jeremy-w We've used [] for optional parameters, and for arrays we used plural form in the name of a parameter.

File: sections/builder.js [hunk 1]
```diff
@@ -283,15 +283,21 @@ export default [
   {
     type: "method",
     method: "with",
-    example: ".with(alias, function|raw)",
-    description: "Add a \"with\" clause to the query. \"With\" clauses are supported by PostgreSQL, Oracle, SQLite3 and MSSQL.",
+    example: ".with(alias, [*columns], callback|builder|raw)",
```
Comment (kibertoad) [inline sections/builder.js:286]: just `columns` should be fine, if it is a mandatory array param

Comment (kibertoad) [inline sections/builder.js:286]: btw, wouldn't it be more correct that third parameter is optional, and second one could be either `columns`, or whatever third parameter could be as well?

Comment (jeremy-w) [reply sections/builder.js:286]: Both with(alias, query) and with(alias, columns, query) are valid. It sounds like [columns] would be the way to write it though, without the star?

Comment (jeremy-w) [reply sections/builder.js:286]: I think it’s equally correct, but I worry it obscures the intended behavior due to the complexity of the type signature:

    with(alias, columns | (qb|callback|raw), (qb|callback|raw) | undefined)

The simplest way to document this IMO would be as two functions with different arities - which is how the TypeScript types are written, in fact. I don’t know that the documentation generator supports that notion of overloaded functions directly, though?

Comment (kibertoad) [reply sections/builder.js:286]: yup, fair

Comment (kibertoad) [reply sections/builder.js:286]: I doubt it, so your original format might be better

**Comment by jeremy-w** [comment]
Updated to use the `[columns]` styling of the optional column list.

**Review by kibertoad — APPROVED** [review]

**Pull request merged**